{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "80c8da0c",
   "metadata": {},
   "source": [
    "# CONSENSUS MOTIF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bad2d40",
   "metadata": {},
   "source": [
    "we can form a consensus string, denoted Consensus(Motifs), from the most popular nucleotides in each column of the motif matrix (ties are broken arbitrarily). If we select Motifs correctly from the collection of upstream regions, then Consensus(Motifs) provides a candidate regulatory motif for these regions. For example, as shown below, the consensus string for the NF-ÎºB binding sites is \"TCGGGGATTTCC\"."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "253db04a",
   "metadata": {},
   "source": [
    "We can implement Consensus(Motifs) using Count(Motifs) as a subroutine. To do so, note that the j-th symbol of this consensus string is equal to the symbol corresponding to a maximum element in column j of Count(Motifs).\n",
    "\n",
    "To implement this idea in Python, first set k equal to the length of Motifs[0] (as we did before) and count equal to the count matrix of Motifs.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2521cac5",
   "metadata": {},
   "source": [
    "```python\n",
    "k = len(Motifs[0])\n",
    "    count = Count(Motifs)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d32660fb",
   "metadata": {},
   "source": [
    "Then, initialize an empty consensus string, and range through each column of the count matrix, adding the maximum element from column j at step j."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f886bc8",
   "metadata": {},
   "source": [
    "```python\n",
    "consensus = \"\"\n",
    "    for j in range(k):\n",
    "        m = 0\n",
    "        frequentSymbol = \"\"\n",
    "        for symbol in \"ACGT\":\n",
    "            if count[symbol][j] > m:\n",
    "                m = count[symbol][j]\n",
    "                frequentSymbol = symbol\n",
    "        consensus += frequentSymbol\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb3bb6be",
   "metadata": {},
   "source": [
    "- Sample Input:\n",
    "```python\n",
    "AACGTA\n",
    "CCCGTT\n",
    "CACCTT\n",
    "GGATTA\n",
    "TTCCGG\n",
    "```\n",
    "- Sample Output:\n",
    "\n",
    "CACCTA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e776b0fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CACCTA'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Insert your Count(Motifs) function here.\n",
    "def Count(Motifs):\n",
    "    count = {}\n",
    "    k = len(Motifs[0])\n",
    "    for symbol in \"ATGC\":\n",
    "        count[symbol] = []\n",
    "        for j in range(k):\n",
    "            count[symbol].append(0)\n",
    "    t = len(Motifs)\n",
    "    for i in range(t):\n",
    "        for j in range(k):\n",
    "            symbol = Motifs[i][j]\n",
    "            count[symbol][j]+= 1\n",
    "    return count\n",
    "# Input:  A set of kmers Motifs\n",
    "# Output: A consensus string of Motifs.\n",
    "def Consensus(Motifs):\n",
    "    k = len(Motifs[0])\n",
    "    count = Count(Motifs)\n",
    "    consensus = \"\"\n",
    "    for j in range(k):\n",
    "        m = 0\n",
    "        frequentSymbol = \"\"\n",
    "        for symbol in \"ACGT\":\n",
    "            if count[symbol][j] > m:\n",
    "                m = count[symbol][j]\n",
    "                frequentSymbol = symbol\n",
    "        consensus += frequentSymbol\n",
    "    return consensus\n",
    "\n",
    "Consensus([\"AACGTA\",\"CCCGTT\",\"CACCTT\",\"GGATTA\",\"TTCCGG\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d4f1f73",
   "metadata": {},
   "source": [
    "### In simpler terms, it's the nucleotide that appears most frequently at each position in the motif matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "075c4c64",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count(motif):\n",
    "    k = len(motif[0])\n",
    "    count = {}\n",
    "    for symbol in \"ATGC\":\n",
    "        count[symbol] = []\n",
    "        for j in range(k):\n",
    "            count[symbol].append(0)\n",
    "    t = len(motif)\n",
    "    for i in range(t):\n",
    "        for j in range(k):\n",
    "            symbol = motif[i][j]\n",
    "            count[symbol][j] += 1\n",
    "    return count\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "397bdd8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def consensus(motif):\n",
    "    k = len(motif)\n",
    "    count = count(motif)\n",
    "    consensus = \"\"\n",
    "    for j in range(k):\n",
    "        m = 0\n",
    "        frequentsymbol = \"\"\n",
    "        for symbol in \"ATGC\":\n",
    "            if count[symbol][j] > m:\n",
    "                m = count[symbol][j]\n",
    "                frequentsymbol = symbol\n",
    "        consensus = consensus + frequentsymbol\n",
    "    return consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1aced012",
   "metadata": {},
   "outputs": [],
   "source": [
    "def score(motif):\n",
    "    k = len(motif[0])\n",
    "    t = len(motif)\n",
    "    score = 0\n",
    "    consensus = Consensus(motif)\n",
    "    for i in range(t):\n",
    "        for j in range(k):\n",
    "            if motif[i][j] != consensus[j] :\n",
    "                score += 1\n",
    "    return score\n",
    "def consensus(motif):\n",
    "    k = len(motif[0])\n",
    "    consensus = \"\"\n",
    "    count = count(motif)\n",
    "    for j in range(k):\n",
    "        m = 0 \n",
    "        frequentsymbol = \"\"\n",
    "        for symbol in \"ATGC\":\n",
    "            if count[symbol][j] > m :\n",
    "                m = count[symbol]\n",
    "                frequentsymbol = symbol\n",
    "        consensus = consensus + frequentsymbol\n",
    "    return consensus\n",
    "def count(motif):\n",
    "    k = len(motif[0])\n",
    "    count={}\n",
    "    for symbol in \"ATGC\":\n",
    "        count[symbol] = []\n",
    "        for j in range(k):\n",
    "            count[symbol][j].append(0)\n",
    "    t = len(motif)\n",
    "    for i in range(t):\n",
    "        for j in range(k):\n",
    "            symbol = motif[i][j]\n",
    "            count[symbol][j] += 1\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a25ba42e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pr(genome,profile):\n",
    "    n = len(genome)\n",
    "    p = 1\n",
    "    for i in range(n):\n",
    "        p = p * profile[genome[i]][i]\n",
    "    return p"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
